### ISSUE TITLE
False positive CT_CONSTRUCTOR_THROW on abstract and abstract sealed classes

### ISSUE THREAD

**Issue by garydgregory** [spotbugs/spotbugs#4259]
It seems to me the 2 examples below are `CT_CONSTRUCTOR_THROW ` false positives since `Bar` is the only type to instantiate and it is `final`.

```java
import java.util.Objects;

import com.rs.seagull.httpmonitor.util.AbstractSealedFoo.Bar;

public sealed abstract class AbstractSealedFoo permits Bar {

    public static final class Bar extends AbstractSealedFoo {

        public Bar(final Object o) {
            super(o);
        }
    }

    private final Object o;
    
    public AbstractSealedFoo(final Object o) {
        this.o = Objects.requireNonNull(o, "o");
    }

    Object getO() {
        return o;
    }
}
```

```java
import java.util.Objects;

public abstract class AbstractFoo {

    public static final class Bar extends AbstractFoo {

        public Bar(final Object o) {
            super(o);
        }
    }

    private final Object o;
    
    public AbstractFoo(final Object o) {
        this.o = Objects.requireNonNull(o, "o");
    }

    Object getO() {
        return o;
    }
}
```

Using:
```
openjdk version "17.0.17" 2025-10-21
OpenJDK Runtime Environment Temurin-17.0.17+10 (build 17.0.17+10)
OpenJDK 64-Bit Server VM Temurin-17.0.17+10 (build 17.0.17+10, mixed mode, sharing)
```
and also:
```
Apache Maven 3.9.16 (2bdd9fddda4b155ebf8000e807eb73fd829a51d5)
Maven home: C:\java\apache-maven-3.9.16
Java version: 17.0.17, vendor: Eclipse Adoptium, runtime: C:\Program Files\Eclipse Adoptium\jdk-17.0.17.10-hotspot
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 11", version: "10.0", arch: "amd64", family: "windows"
```

With spotbugs-maven-plugin 4.10.4.0 and spotbugs 4.10.4

**Comment by JuditKnoll** [comment]
Thank you for the report.
The first example is really a false positive, I'll create a PR fixing it.
The second example looks like a true positive to me. Since `AbstractFoo` is public, not sealed and the constructor is public, there could be other classes descending from it and calling the constructor.

**Comment by garydgregory** [comment]
Hi @JuditKnoll 
Thank you for the update. I think there is a third case I didn't list which is like the second (non-sealed) but with package private classes, which we then can tell whether there are other subclasses in that package. WDYT?

**Comment by JuditKnoll** [comment]
> Hi [@JuditKnoll](https://github.com/JuditKnoll) Thank you for the update. I think there is a third case I didn't list which is like the second (non-sealed) but with package private classes, which we then can tell whether there are other subclasses in that package. WDYT?

Using a package private non-sealed non-final class would only be enough to prevent malicious subclassing, if both of these conditions are met:
- it has no public non-sealed non-final descendants in the package,
- there is a module defined with the package not exported.

Currently class visibility and module check is not implemented in the detector, so there is still a `CT_CONSTRUCTOR_THROW` hit in this edge case as well.

On the topic of the first example, sealed class and [permitted sub class](https://github.com/apache/commons-bcel/blob/master/src/main/java/org/apache/bcel/classfile/PermittedSubclasses.java) handling is introduced in commons-bcel 6.13.0, which is not released yet. For a proper fix that would be required, so I ask for your patience until then. On the other hand, do you have perhaps a hint about when the next release of commons-bcel is expected?

**Comment by garydgregory** [comment]
@JuditKnoll 
I'll review pending bug reports and see if I can create a release candidate over the weekend.

**Comment by garydgregory** [comment]
@JuditKnoll 
I released Commons BCEL 6.13.0 to Maven Central.

**Comment by garydgregory** [comment]
Hi @JuditKnoll 
Any thoughts on when we should expect a fix? If it's soon, I'll hold up releasing `commons-parent`. Ty!

**Comment by JuditKnoll** [comment]
For fixing the first example, handling sealed classes I opened https://github.com/spotbugs/spotbugs/pull/4322. That one is waiting for review. For the next release, [4.10.5](https://github.com/spotbugs/spotbugs/milestone/37) milestone currently September 30, 2026 is set as a due date.
The second example is a true positive.
The third one requires being able to check module files when visiting the class. I looked into this, and unfortunately, that's outside of SpotBugs capabilities.

**Comment by garydgregory** [comment]
Thank you for the detailed update. I'll wait for 4.10.5 to release commons-parent 106.

**Comment by JuditKnoll** [comment]
Since https://github.com/spotbugs/spotbugs/pull/4322 fixing the first example got merged to master, the second example is a TP, and the third one is outside of SpotBugs capabilities, I'm closing this issue.
